Dip/Dip/Dip.swift

```diff
@@ -77,24 +97,41 @@ public class DependencyContainer {
   Register a Void->T factory associated with optional tag.
   
   - parameter tag: The arbitrary tag to associate this factory with when registering with that protocol. Pass `nil` to associate with any tag. Default value is `nil`.
+  - parameter scope: scope to use for this compone
   - parameter factory: The factory to register, with return type of protocol you want to register it for
+  - returns: definition created for provided type and factory
+  
+  - note: You must cast the factory return type to the protocol you want to register it for.
+  Inside factory block if you need to reference container use it as `unowned` to avoid retain cycle.
   
-  - note: You must cast the factory return type to the protocol you want to register it with (e.g `MyClass() as MyAPI`)
+  **Example**:
+  ```swift
+  container.register { ServiceImp() as Service }
+  container.register(tag: "service") { ServiceImp() as Service }
+  container.register(.ObjectGraph) { ServiceImp() as Service }
+  container.register { [unowned container] 
+    ClientImp(service: container.resolve() as Service) as Client 
+  }
+  ```
   */
-  public func register<T>(tag tag: Tag? = nil, factory: ()->T) -> DefinitionOf<T> {
-    return register(tag: tag, factory: factory, scope: .Prototype) as DefinitionOf<T>
+  public func register<T>(tag tag: Tag? = nil, _ scope: ComponentScope = .Prototype, factory: ()->T) -> DefinitionOf<T, ()->T> {
+    return registerFactory(tag: tag, scope: scope, factory: factory)
   }
   
   /**
    Register a Singleton instance associated with optional tag.
    
-   - parameter tag: The arbitrary tag to associate this instance with when registering with that protocol. `nil` to associate with any tag.
+   - parameter tag: The arbitrary tag to associate this instance with when registering with that protocol. 
+   Pass `nil` to associate with any tag.
    - parameter instance: The instance to register, with return type of protocol you want to register it for
    
    - note: You must cast the instance to the protocol you want to register it with (e.g `MyClass() as MyAPI`)
+   
+   **Deprecated**: Use `register(.Singleton){}` method instead to define singleton scope.
    */
-  public func register<T>(tag tag: Tag? = nil, @autoclosure(escaping) instance factory: ()->T) -> DefinitionOf<T> {
-    return register(tag: tag, factory: { factory() }, scope: .Singleton)
+  @available(*, deprecated, message="Use `register(.Singleton){}` method instead to define singleton scope.")
```

Comment:
:cry: I liked the use of `@autoclosure` here, that's not every day we have a reason to use it :stuck_out_tongue_winking_eye: 

Too bad I didn't see that depreciation before, it would have made sense to include it in 3.0.0 so that we could remove the method altogether in the next 4.0.0.

If the next release changes the major version (if the API of your next PRs change from the API of 3.0.0 and isn't backwards compatible anymore) maybe we will do a 3.1.0 that just adds this depreciation warning before publishing 4.0.0 so we can completely remove that method in 4.0.0 then?